### ISSUE TITLE
st_transform() returns empty results if anything in a multipoint is problematic

### ISSUE THREAD

**Issue by dankelley** [r-spatial/sf#1223]
Hello.  First, thanks for "sf", which is a great package. I am trying to convert the "oce" package to use "sf" for its map-projection work, because `rgdal::project()`, which "oce" presently uses, seems to be slated for removal, and `rgdal::transform()` returns nothing if a single point in a vector of points is uninvertable (see https://github.com/dankelley/oce/issues/1629#issue-536962648 for some discussion of the oce context).

However, I am finding that "sf" also has this problem of a single un-invertable point causing a problem.  The code given below demonstrates (see comments in the code). As a test, I removed the check for error condition 6 from https://github.com/r-spatial/sf/blob/abbd0a9254ebf12b572d4de5f3f6db362626aaef/src/gdal.cpp#L402 and then I got back a sequence of points.  However, this is not the ideal condition. It would seem to be preferable to set the problematic point to NA or Inf, as `rgdal::project()` does.

I have not examined the gdal (or sf) code in sufficient detail to suggest how to accomplish what `rgdal::project()` does.  I would love to be able to stick with `sf`, if I can, because I am under the impression that `sf::transform()` is destined to live on there for a while, whereas `rgdal::project()` is going to disappear at some point, crippling map projection in oce.

PS. I know that this is an overly long issue report. Also I apologize for using incorrect words for the object types (polygon, multipoint, etc).

```R
# Demonstrate an issue about offworld inverse projection with multipoint
# data. If a multipoint contains *any* data that ar sufficiently far from
# the projected earth, then sf::st_transform returns an object containing
# no data.  It would be helpful if only those problematic points within
# the multipoint object were flagged (e.g. with NA or Inf values), as is
# done with rgdal::project().
#
# 2019-12-20 Dan Kelley, Dalhousie University (kelley.dan@gmail.com)
#
# Test platform: macos catalina (beta), R-devel downloaded 2019-12-19,
# sf either from CRAN or from github (commit
# abbd0a9254ebf12b572d4de5f3f6db362626aaef)

library(sf)

# 1. Check that forward and inverse transforms work for some equatorial
# points that are near the "edge" of the projected earth.
lon <- c(177, 178, 179)
lat <- c(  0,   0,   0)
lonlat <- unname(cbind(lon, lat))
lonlatSF <- sf::st_sfc(sf::st_multipoint(lonlat), crs=sf::st_crs("+proj=longlat"))
# Forward: no errors
xySF <- sf::st_transform(lonlatSF, sf::st_crs("+proj=robin"), check=FALSE, partial=TRUE)
# Inverse: recovers input
xy <- unname(sf::st_coordinates(xySF))[, 1:2]
lonlat2SF <- sf::st_coordinates(sf::st_transform(xySF, crs=sf::st_crs("+proj=longlat")))
lonlat2 <- unname(lonlat2SF)[, 1:2]
stopifnot(all.equal(lonlat, lonlat2))

# 2. Demonstrate the problem, by moving one of our three points
# progressively farther "off the edge" of the projected earth.
# This produces a lot of output ... but the pattern is visible
# at a glance: on my macos (Catalina) machine with R-devel,
# this happens between 8600km and 8700 km. Note that this
# is so for any combination of check/partial values.
check <- TRUE
partial <- TRUE
for (delta in 1000e3*seq(8, 9, 0.1)) {
    DELTA <- matrix(c(0, delta, 0, 0, delta, 0), ncol=2)
    tmp <- sf::st_sfc(sf::st_multipoint(xy+DELTA), crs=sf::st_crs("+proj=robin"))
    cat("\ndelta=", delta/1000, "km ... does this yield a null result?\n")
    print(sf::st_coordinates(sf::st_transform(tmp, crs=sf::st_crs("+proj=longlat"),
                                              check=check, partial=partial)))
}
```

**Comment by rsbivand** [comment]
The deprecation of `rgdal::project()` is a consequence of my reading of https://github.com/OSGeo/PROJ/blob/master/src/fwd.cpp, where the PRøJ function `proj_trans()` gets called anyway per coordinate. This means that the user will often be faced with datum transformation anyway, and with the need to choose between pipeline proposals of different accuracy if I understand correctly. It will be a different world, I'm afraid.

**Comment by dankelley** [comment]
@rsbivand -- Thanks for the comment.  I made up a test case and attached it as `sf02_R.txt` (since github does not permit .R attachments) and the associated output on my macos machine is attached as `sf02_out.txt`.

At the very end, I have denoted with "NOTE 2" my main problem, which is that `rgdal::transform` returns no value the whole set of points, if any one point is problematic.  

Earlier on, I have denoted as "NOTE 1" an issue that I had noticed before, related to differing `rgdal::project()` treatment of off-world points by different projections.

The various `stopifnot` statements are mainly to guard against typos and other errors of my own coding ... they can be taken like `assert()` calls in C, whose purpose is just to state the expectation of the coder.

I hope my code is not too wordy, and that I have explained myself adequately.

Thanks again!

--  Dan

[sf02_R.txt](https://github.com/r-spatial/sf/files/3991378/sf02_R.txt)

[sf02_out.txt](https://github.com/r-spatial/sf/files/3991380/sf02_out.txt)

**Comment by edzer** [comment]
@dankelley in your for loop, have you tried using
```r
print(sf::sf_project("+proj=robin", "+proj=longlat", st_coordinates(tmp)[,1:2]))
```
It generates `Inf` values for bad points, and warnings if this happens. I'm about to change this into `NA` values and add an option to not generate warnings.

**Comment by dankelley** [comment]
@edzer Such a change to `sf::sf_project` would be very welcome, thanks.  Somehow, I had thought that the preferred function was `sf::st_transform`, which is why I was trying that.  (To be honest, I am a bit confused by the variety of approaches, and particularly their potential in terms of working with future versions of the underlying PROJ code.

**Comment by rsbivand** [comment]
Dan,

Confused is as good as it gets at the moment. Despair is a valid option,
but we're trying to avoid that!

From an airport,

Roger

Roger Bivand
Falsensvei 32
5063 Bergen

lør. 21. des. 2019, 20.16 skrev Dan Kelley <notifications@github.com>:

**Comment by edzer** [comment]
The difference is this: `st_transform` uses the transform functions in GDAL, which operate on complete feature geometry structures, and use PROJ underneath. But it does more: it refuses to handle transformations that are not invertible, and raises errors when one or more of the points cannot be transformed (which would otherwise lead to an geometry with one or more invalid nodes). `sf_project` does a low-level transformation directly using PROJ, skipping GDAL altogether, point by point. 

(`lwgeom::st_transform_proj` does something similar, but for entire feature geometries, also skipping GDAL, but will also not solve the problem of invalid geometries on return)

**Comment by rsbivand** [comment]
I'll see if I can backport from st_project() to rgdal::project(), to try to
accommodate users of raster.

Roger Bivand
Falsensvei 32
5063 Bergen

lør. 21. des. 2019, 20.58 skrev Edzer Pebesma <notifications@github.com>:

**Comment by edzer** [comment]
Given that #1228 got closed, can we also close this one?

**Comment by dankelley** [comment]
Yes, I'll close it.  Many thanks!  Dan.
